src/Uno.UI/UI/Xaml/UIElement.Layout.netstd.cs

```diff
@@ -131,13 +237,91 @@ private void MeasureVisualTreeRoot(Size availableSize)
 
 		private void DoMeasure(Size availableSize)
 		{
-			InvalidateArrange();
+			var isDirty =
+				(availableSize != LastAvailableSize)
+				|| IsMeasureDirty
+				|| !GetLayoutFlag(LayoutFlag.FirstMeasureDone);
+
+			if (!isDirty && !IsMeasureDirtyPath)
+			{
+				return; // Nothing to do
+			}
+
+			if (!GetLayoutFlag(LayoutFlag.FirstMeasureDone))
+			{
+				SetLayoutFlag(LayoutFlag.FirstMeasureDone);
+				isDirty = true;
+			}
 
-			// We must reset the flag **BEFORE** doing the actual measure, so the elements are able to re-invalidate themselves
-			_isMeasureValid = true;
+			var remainingTries = MaxLayoutIterations;
 
-			MeasureCore(availableSize);
-			LayoutInformation.SetAvailableSize(this, availableSize);
+			while (--remainingTries > 0)
+			{
+				if (isDirty)
+				{
+					// We must reset the flag **BEFORE** doing the actual measure, so the elements are able to re-invalidate themselves
+
+					ClearLayoutFlags(LayoutFlag.MeasureDirty | LayoutFlag.MeasureDirtyPath);
+
+					// The dirty flag is explicitly set on this element
+#if DEBUG
+					try
+#endif
+					{
+						MeasureCore(availableSize);
+						InvalidateArrange();
+					}
+#if DEBUG
+					catch (Exception ex)
+					{
+						_log.Error($"Error measuring {this}", ex);
+					}
+					finally
+#endif
+					{
+						LayoutInformation.SetAvailableSize(this, availableSize);
+					}
+
+					break;
+				}
+
+				if (IsMeasureDirtyPath)
+				{
+					ClearLayoutFlags(LayoutFlag.MeasureDirtyPath);
+
+					// The dirty flag is set on one of the descendents:
+					// it will bypass the current element's MeasureOverride()
+					// since it shouldn't produce a different result and it's
+					// just a waste of precious CPU time to call it.
+					var children = GetChildren();
+
+					foreach (var child in children)
```

Comment:
Use explicit enumeration instead, use the same comment:
https://github.com/unoplatform/uno/blob/6c828d2bb8e7e61bfc04868bd60648c9382228a8/src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs#L1086-L1089